src/blockchain_db/sqlite/db_sqlite.cpp

```diff
@@ -356,37 +426,55 @@ void BlockchainSQLite::update_height(uint64_t new_height) {
     prepared_exec("UPDATE batch_db_info SET height = ?", static_cast<int64_t>(height));
 }
 
-void BlockchainSQLite::blockchain_detached(uint64_t new_height) {
-    if (height < new_height)
-        return;
-    int64_t revert_to_height = new_height - 1;
-    auto maybe_prev_interval = prepared_maybe_get<int64_t>(
-            "SELECT MAX(archive_height) FROM batched_payments_accrued_archive "
-            "WHERE archive_height <= ?",
-            revert_to_height);
-
-    if (!maybe_prev_interval) {
-        auto fork_height =
-                cryptonote::hard_fork_begins(m_nettype, hf::hf19_reward_batching).value_or(0);
-        reset_database();
-        update_height(fork_height);
-        return;
-    }
-    const auto prev_interval = *maybe_prev_interval;
+void BlockchainSQLite::blockchain_detached(AccruedTableType history, uint64_t new_height) {
+    const auto& netconf = get_config(m_nettype);
 
-    db.exec(fmt::format(
-            R"(
-      DELETE FROM batched_payments_raw WHERE height_paid > {0};
+    // NOTE: Execute detach
+    std::string detach_label = "";
+    int rows_restored = 0;
+    int rows_removed = batch_payments_accrued_row_count(AccruedTableType::Nil, nullptr);
+    switch (history) {
+        case AccruedTableType::Nil: {
+            reset_database();
+            detach_label = " (via reset)";
+        } break;
+
+        default: {
+            std::string history_table = "batched_payments_accrued_{}"_format(
+                    history == AccruedTableType::Archive ? "archive" : "recent");
+            rows_restored = batch_payments_accrued_row_count(history, &new_height);
+
+            db.exec(R"(DELETE FROM batched_payments_raw WHERE height_paid > {0};
+                       DELETE FROM batched_payments_accrued;
+                       DELETE FROM batched_payments_accrued_recent WHERE height > {0};
+                       DELETE FROM batched_payments_accrued_archive WHERE height > {0};
+
+                       INSERT INTO batched_payments_accrued
+                       SELECT address, amount, payout_offset
+                       FROM {1} WHERE height = {0};
+              )"_format(new_height, history_table));
+
+            detach_label = history == AccruedTableType::Archive ? " (from archive history)"
+                                                                : " (from recent history)";
+        } break;
+    }
 
-      DELETE FROM batched_payments_accrued;
+    // NOTE: When we detach we are rewinding to an old state. The accrued table
```

Comment:
I'm not sure this is true.  I think because of the trigger `batch_payments_delete_empty` that an archive at an older height can have more rows, as someone fully unstaking and cashing out will have their addresses' rows removed.

Reply:
Oh that's interesting. I didn't notice that trigger. Though if someone fully unstakes and cashes out, that is an additive operation to their row. Their exit stake gets added into their row's `amount` counter. 

But, now that I've removed the popping of blocks from the DB which would subtract rewards on a block-by-block basis (because popping breaks when you remove a delayed payment), there's no operation you can do on the DB to subtract. 

On detach you delete all the rows and restore all the old rows, and to be present in a DB snapshot you must have a non-zero rewards amount so you can't ever encounter a row with 0 rewards for this trigger to execute?

If that's the case, then I'll push a commit to remove the trigger if you don't see an issue with that.

Reply:
by "cashing out" I meant claiming the entirety of their rewards, forgetting that we indeed no longer subtract from that table after HF21.  Up until that point, I suppose when an address' batch payment goes through its row would zero out, so there's maybe a window of time where we have to care.

Reply:
I have removed this check